### PR TITLE
stream redirector wip

### PR DESCRIPTION
**Pull request by 0xdavinchee**

### CODE REVIEW

File: packages/ethereum-contracts/contracts/superfluid/Superfluid.sol [hunk 1]
```diff
@@ -514,11 +517,11 @@ contract Superfluid is
         returns (bytes memory appCtx)
     {
         Context memory context = decodeCtx(ctx);
-        if (isApp(ISuperApp(context.msgSender))) {
+        if (isApp(ISuperApp(context.msgSender)) && context.appCallbackLevel >= MAX_APP_CALLBACK_LEVEL) {
```
Comment (0xdavinchee) [inline packages/ethereum-contracts/contracts/superfluid/Superfluid.sol:520]: Logic Change A

File: packages/ethereum-contracts/contracts/superfluid/Superfluid.sol [hunk 2]
```diff
@@ -592,8 +595,8 @@ contract Superfluid is
         bytes4 agreementSelector = CallUtils.parseSelector(callData);
 
         //Build context data
-        bytes memory  ctx = _updateContext(Context({
```
Comment (0xdavinchee) [inline packages/ethereum-contracts/contracts/superfluid/Superfluid.sol:595]: Logic Change B

File: packages/ethereum-contracts/contracts/superfluid/Superfluid.sol [hunk 3]
```diff
@@ -638,7 +641,7 @@ contract Superfluid is
     {
         //Build context data
         bytes memory ctx = _updateContext(Context({
-            appLevel: isApp(ISuperApp(msgSender)) ? 1 : 0,
```
Comment (0xdavinchee) [inline packages/ethereum-contracts/contracts/superfluid/Superfluid.sol:641]: Logic Change C

File: packages/ethereum-contracts/contracts/superfluid/Superfluid.sol [hunk 4]
```diff
@@ -514,11 +517,11 @@ contract Superfluid is
         returns (bytes memory appCtx)
     {
         Context memory context = decodeCtx(ctx);
-        if (isApp(ISuperApp(context.msgSender))) {
```
Comment (0xdavinchee) [inline packages/ethereum-contracts/contracts/superfluid/Superfluid.sol:517]: Logic Change A

**Comment by 0xdavinchee** [comment]
Reason for foundry change: https://github.com/foundry-rs/foundry/issues/2572

**Comment by hellwolf** [comment]
@d10r please have a look of this.

File: packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperfluid.sol [hunk 5]
```diff
@@ -69,7 +69,7 @@ interface ISuperfluid {
      * @dev Register a new agreement class to the system
      * @param agreementClassLogic Initial agreement class code
      *
-     * @custom:modifiers 
+     * @custom:modifiers
```
Comment (d10r) [inline packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperfluid.sol]: I'm afraid removing those trailing whitespaces after natspec tags will affect the documentation - last I tried the parser wouldn't recognize tags immediately followed by a newline.
I guess this was done automatically by the IDE. Could we disable removal of trailing whitespaces in comment sections?

File: packages/ethereum-contracts/contracts/mocks/StreamRedirector.sol [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+     * Terminates a stream from an _originAccount who has granted this SuperApp ACL permissions
+     * @param _originAccount ACL permissions granter
+     */
+    function stopStreamToSelf(address _originAccount) public {
+        cfaV1.deleteFlowByOperator(_originAccount, address(this), token);
+    }
+
+    /**
+     * Creates a stream from this Super App to another SuperApp.
+     * @param _superApp the targetted super app
```
Comment (d10r) [inline packages/ethereum-contracts/contracts/mocks/StreamRedirector.sol]: typo in "targetted"

File: packages/ethereum-contracts/test/contracts/agreements/ConstantFlowAgreementV1.test.js [hunk 7]
```diff
@@ -5544,5 +5547,563 @@ describe("Using ConstantFlowAgreement v1", function () {
...
+                receiver: "redirector",
+                flowRate: toBN(0),
+                deposit: toBN(0),
+                owedDeposit: toBN(0),
+            });
+
+            // State: No flows
+        });
+
+        it("#4.30 SuperApp with permissions should be able to stream to itself (blue elephant)", async () => {
```
Comment (d10r) [inline packages/ethereum-contracts/test/contracts/agreements/ConstantFlowAgreementV1.test.js:5690]: finally he made it into the code!

File: packages/ethereum-contracts/contracts/superfluid/Superfluid.sol (hunk 4 above)
Comment (d10r) [reply packages/ethereum-contracts/contracts/superfluid/Superfluid.sol:517]: Note to self: not clear why this check can now be skipped if appCallbackLevel < MAX_APP_CALLBACK_LEVEL.

**Review by d10r — CHANGES_REQUESTED** [review]

File: packages/ethereum-contracts/test/contracts/agreements/ConstantFlowAgreementV1.test.js (hunk 7 above)
Comment (0xdavinchee) [reply packages/ethereum-contracts/test/contracts/agreements/ConstantFlowAgreementV1.test.js:5690]: 😉 🐘

File: packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperfluid.sol (hunk 5 above)
Comment (0xdavinchee) [reply packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperfluid.sol]: undid these changes, my IDE automatically does this, but will disable my IDE from doing this

**Review by jtriley2p — DISMISSED** [review]
Callbacks allow arbitrary super tokens, this app can be forcibly jailed.

**Comment by 0xdavinchee** [comment]
@d10r @JoshuaTrujillo15 reviews addressed

**Review by d10r — APPROVED** [review]

**Review by hellwolf — APPROVED** [review]
great refactoring

**Pull request merged**